### ISSUE TITLE
[RFC] Connector error reporting.

### ISSUE THREAD

**Issue by ryzhyk** [feldera/feldera#1108]
The API currently only reports the number of non-critical errors and the last critical error for each input and output endpoint. Non-critical errors, e.g., parsing errors, can only be found in the log.  While we should make Feldera logs viewable in all form factors, we should also expose the errors via the API. To do this, we need to figure out a way to buffer and rate limit errors, e.g., parsing errors can get generated at a high rate when consuming a stream of invalid JSON records, and we don't want to overwhelm the API client with them.

### Option 1

The simplest solution is to add an `errors` field to per-connector stats returned by the `/stats` endpoint. This field can work as a circular buffer that always returns the last N errors.

### Option 2

Add a new API endpoint for errors. This option can give the client additional control over how errors are reported, e.g., the client may specify the last error they have observed and only receive new errors that have occurred since then, or they may want to request a report with selected error types only.

### Option 3

Model errors as Feldera tables. Every pipeline has an additional `error` system table and some predefined views over this table.

```sql
create table 'feldera.error' (
    serial_number bigint,
    error_type varchar,
    endpoint varchar,
    message varchar,
);
```

The table and views can be consumed using regular API for tables (e.g., read through the `/egress` endpoint or sent to a Kafka stream). But we can also provide an additional convenience API for working with errors.

Pros
====

- Potentially more powerful: we will be able to use the Feldera query engine to analyze the error stream. This will become more useful as we add support for ad hoc queries in the future. 

Cons
====

- Complexity: we will need to introduce the concept of system tables and treat them in a special way in the WebConsole and the API, e.g., the error table is read-only.

@Karakatiza666 , @lalithsuresh , @gz

**Comment by mihaibudiu** [comment]
how about runtime errors generated by the pipeline?

**Comment by ryzhyk** [comment]
> how about runtime errors generated by the pipeline?

Any of the three approaches should work for that too.

**Comment by mihaibudiu** [comment]
There is no way we can avoid system tables. The sooner we embrace them, the sooner we can benefit from them. We can put there everything that a database does, including the catalog itself.

**Comment by ryzhyk** [comment]
> Instead of an errors endpoint (option 2), I'd advocate for an events or pipeline/{pipeline_id}/events endpoint that is a stream of events. Events can be scoped per tenant and their API objects (some example events: ProgramCompiled, PipelineDeployed, PipelineFailed... etc).

> I'd say events should largely be scoped to pipeline-manager level events, and is not the the right vehicle for connector/pipeline runtime errors (for the simple reason that you can have N pipelines each processing large data volumes, but we will always only have 1 logical pipeline manager, and the rates don't line up).

So in summary you are suggesting that we need separate mechanisms for events and errors.  Errors should be implemented as systems tables, whereas events should be exposed via REST endpoints.  Is this correct?

What's included in the pipeline manager level events?  More importantly, what is not included (and hence will require a different mechanisms), compile/start/shutdown pipeline makes sense.  What about pause/unpause a pipeline? HTTP ingress/egress request? Backpressure due to full buffers, checkpoints, etc.

**Comment by ryzhyk** [comment]
> There is no way we can avoid system tables. The sooner we embrace them, the sooner we can benefit from them. We can put there everything that a database does, including the catalog itself.

I agree.  It is more work, but the work will need to be done anyway, so why not now.  This will only become practical once we've solved the compilation speed problem, otherwise system tables will make compile times worse.

**Comment by lalithsuresh** [comment]
> So in summary you are suggesting that we need separate mechanisms for events and errors. Errors should be implemented as systems tables, whereas events should be exposed via REST endpoints. Is this correct?

Yes, although the distinction is blurry. See below.

> What's included in the pipeline manager level events? More importantly, what is not included (and hence will require a different mechanisms), compile/start/shutdown pipeline makes sense. What about pause/unpause a pipeline? HTTP ingress/egress request? Backpressure due to full buffers, checkpoints, etc.

My first instinct is to say control plane activity should be events and datapath activity should go into system tables. Control plane activity would be API level events for program/pipeline/connector/service etc as well as their state transitions (which would cover pipeline start/pause and perhaps "started an HTTP egress stream").  Datapath activity would be connector errors, pipeline errors (e.g., divide by zero etc), backpressure etc.

**Comment by lalithsuresh** [comment]
> > What's included in the pipeline manager level events? More importantly, what is not included (and hence will require a different mechanism). Compile/start/shutdown pipeline makes sense as manager-level events. What about pause/unpause a pipeline? HTTP ingress/egress request? Backpressure due to full buffers, checkpoints, etc.
> 
> I guess, the distinction between events and errors is misleading. They are all events. But there are pipeline-level events, which can be high-volume and we want to model them as systems tables. And then there are manager level events, which cannot be attributed to a pipeline.

exactly!

**Comment by lalithsuresh** [comment]
The idea of system pipelines could make sense. Per-tenant system pipelines worry me because if we have a 1000 tenants, we'll have a 1000 pipelines (e.g., on the sandbox). It also means we'll have to create system pipelines on the fly. 

A per-manager system pipeline makes sense, but reading only some tenants' data from one would be tricky. Do we read from the tables directly via HTTP or do we output the events somehow into Postgres?

**Comment by ryzhyk** [comment]
Yes, this is a problem. The reason system tables are nice is because they unify data and metadata management, so they should be accessible via the standard Feldera API.

**Comment by lalithsuresh** [comment]
If we can make system pipelines work, that'll be one more avenue for us to dogfood.
